### PR TITLE
Add charts to variantdata

### PR DESCRIPTION
**Pull request by EddieLF**
Displays charts on the VariantData page with react, `using chart.js` and the wrapper `react-chartjs-2` ~~and `antd` for~~ in tabs. See https://github.com/populationgenomics/variant-curation-portal/pull/61 for the four new fields added recently. 

The Genotype Qualities can take a value from 1-100 and these are binned into bars of width 5. 

The Read Depths can take a value from 1-N and are binned into bars of width 5 also.

The Allele Balances are calculated from the Allelic Depth array `[[REF,ALT], [REF,ALT], ..., [REF,ALT]]`, where each balance is  `AB=ALT/(REF+ALT)`. These take values in the range [0, 1.0] and  are binned into bars of width 0.05.

The null handling means that for older variants without these fields populated, nothing will be shown on the charts.

---

Please see this video for a demo of the new charts: [Google Drive Link to Screen Recording](https://drive.google.com/file/d/1cm9dFiguSWtUW91TrJKSt5JtoOgFeCOc/view?usp=sharing)

---
This is in an effort to recreate the charts provided by GnomAD, such as this one:
gnomad_chart

### CODE REVIEW

**Comment by cassimons** [comment]
LGTM. Can you add a couple of screen grabs showing what the final result looks like?

**Comment by EddieLF** [comment]
@cassimons 

For a random variant:

Genotype qualities: `[99, 99]`
Screenshot 2024-01-17 at 9 18 29 am

Read Depths: `[76, 52]`
Screenshot 2024-01-17 at 9 18 35 am

Allelic Balances: `[0.37, 0.56]` (from Allelic Depths: `[[48, 28], [23, 29]]`
Screenshot 2024-01-17 at 9 18 43 am


---

For a random variant with no values for `GQ_all`, `DP_all`, `AD_all`, the charts are empty:
Screenshot 2024-01-17 at 9 22 35 am

File: package.json [hunk 1]
```diff
@@ -40,13 +40,16 @@
     "webpack-dev-server": "^4.15.1"
   },
   "dependencies": {
+    "antd": "^5.12.8",
```
Comment (illusional) [inline package.json:43]: I think the tabs can be done with [semantic-ui](https://semantic-ui.com/modules/tab.html) which is already added to the project.

File: assets/components/pages/project/curate/VariantData.js [hunk 2]
```diff
@@ -187,7 +422,7 @@ class VariantData extends React.Component {
         </List.Item>
         <List.Item>
           <strong>Allelic Depths:</strong>{" "}
-          {variant.AD ? variant.AD.map((depths) => `(${depths.join(", ")})`).join(", ") : "N/A"}
+          {(variant.AD ?? []).map((depths) => `(${depths.join(", ")})`).join(", ") || "N/A"}
```
Comment (illusional) [inline assets/components/pages/project/curate/VariantData.js:425]: This optional coalescing is quite modern (>2020 browsers) - would be good to check the compiled bundle to check if this is being transpiled away.

File: assets/components/pages/project/curate/VariantData.js [hunk 3]
```diff
@@ -1,6 +1,11 @@
 import PropTypes from "prop-types";
 import React from "react";
 import { Button, List } from "semantic-ui-react";
+import { Chart, registerables } from "chart.js";
+import { Bar } from "react-chartjs-2";
```
Comment (illusional) [inline assets/components/pages/project/curate/VariantData.js:5]: Can you explain what Chart.register(...registerables) means here?
`chart.js` does have [bar chart functionality](https://www.chartjs.org/docs/latest/charts/bar.html), can we do this with fewer new dependencies?

File: assets/components/pages/project/curate/VariantData.js [hunk 4]
```diff
@@ -147,10 +152,240 @@ class VariantData extends React.Component {
...
+      // Handle the case where GQ_all is null
+      return { data: { labels: [], datasets: [] }, options };
+    }
+
+    // Calculate the frequency of each GQ_all value
+    const gqFrequencies = variant.GQ_all.reduce((acc, value) => {
+      const bin = Math.floor(value / 5) * 5; // Bin values in groups of 5
+      acc[bin] = (acc[bin] || 0) + 1;
+      return acc;
+    }, {});
```
Comment (illusional) [inline assets/components/pages/project/curate/VariantData.js:199]: This is a case where the mdn web docs suggest against using `Array.reduce`, instead opting for a `for ... loop`. `.reduce` is usually used to mimic immutable data practices in functional programming, but this is mutated `acc` on each iteration [reduce#when_to_not_use_reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce#when_to_not_use_reduce).

(Ditto for below too)

File: assets/components/pages/project/curate/VariantData.js [hunk 5]
```diff
@@ -147,10 +152,240 @@ class VariantData extends React.Component {
     showAll: false,
   };
 
+  getGQFrequencies() {
```
Comment (illusional) [inline assets/components/pages/project/curate/VariantData.js:155]: These methods are very similar, would you consider creating a method like `getPreparedBinnedData(data, ...other config options)`

**Review by illusional — CHANGES_REQUESTED** [review]

File: assets/components/pages/project/curate/VariantData.js (hunk 3 above)
Comment (EddieLF) [reply assets/components/pages/project/curate/VariantData.js]: @illusional thanks, I've updated the PR to only use chart.js for now.

The `Chart.register(...registerables)` is for the scales used in the charts. The scale type in the options is set to `"linear"`, and without the `register` call the page will not display:
Screenshot 2024-01-18 at 2 47 18 pm

I may be misunderstanding the necessity of the `scales.type` option, but when I remove the `type: "linear"` line I get the same error page but with `"categorical"` in place of `"linear"` in the error (i.e `"categorical" is not a registered scale`).

Comment (illusional) [reply assets/components/pages/project/curate/VariantData.js]: Yeah fair enough, I had no clue, could you leave a comment with this information?
It looks like it's due to charts.js being tree-shakeable (only bundles the used components).

You could probably explicitly import some functionality, but easier to register it, here's a couple of links:

- https://www.chartjs.org/docs/next/getting-started/integration.html#bundlers-webpack-rollup-etc
- https://stackoverflow.com/a/67240031

File: assets/components/pages/project/curate/VariantData.js [hunk 6]
```diff
@@ -119,6 +122,149 @@ TagsList.propTypes = {
   ).isRequired,
 };
 
+const BarChart = ({ data, options }) => {
+  const chartRef = useRef(null);
+  const chartInstance = useRef(null);
+
+  useEffect(() => {
+    if (chartRef && chartRef.current) {
+      chartInstance.current = new Chart(chartRef.current, {
```
Comment (illusional) [inline assets/components/pages/project/curate/VariantData.js:131]: Oh, I think I misunderstood, the reactjs-charts2 was a React wrapper of the charts.js library, with this information I think your original approach was actually better.

Sorry, I didn't realise that charts.js was going to have anti-react patterning.

File: assets/components/pages/project/curate/VariantData.js (hunk 2 above)
Comment (EddieLF) [reply assets/components/pages/project/curate/VariantData.js:319]: In the compiled bundle, this line is compiled as 
`
t.createElement("strong",null,"Allelic Depths:")," ",(null!==(o=u.AD)&&void 0!==o?o:[]).map((function(e){return"(".concat(e.join(", "),")")})).join(", ")||"N/A")
`

Comment (illusional) [reply assets/components/pages/project/curate/VariantData.js:319]: Dope! Thanks for checking :)

**Comment by cassimons** [comment]
I am happy with this, @illusional can you pls approve if/when you give the 👍 ?

**Review by illusional — APPROVED** [review]

**Pull request merged**